subsys/bluetooth/services/hids_c.c

```diff
@@ -0,0 +1,851 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <kernel.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/hids_c.h>
+
+#define LOG_LEVEL CONFIG_BT_GATT_HIDS_C_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(hids_c);
+
+/* Suspend value for Control Point  */
+#define BT_GATT_HIDS_C_CP_SUSPEND      0x00
+/* Exit suspend value for Control Point */
+#define BT_GATT_HIDS_C_CP_EXIT_SUSPEND 0x01
+
+/* Memory slab used for reports */
+K_MEM_SLAB_DEFINE(bt_gatt_hids_c_reports_mem,
+		  sizeof(struct bt_gatt_hids_c_rep_info),
+		  CONFIG_BT_GATT_HIDS_REPORTS_MAX,
+		  4);
+
+
+/**
+ * @brief Allocate memory for report structure
+ *
+ * Function allocates memory for report structure.
+ * The structure would be cleared.
+ *
+ * @return The pointer to the structure or NULL if no memory is available.
+ *
+ * @sa rep_new
+ */
+static struct bt_gatt_hids_c_rep_info *rep_alloc(void)
+{
+	struct bt_gatt_hids_c_rep_info *rep;
+	if (k_mem_slab_alloc(&bt_gatt_hids_c_reports_mem,
+			     (void**)&rep, K_NO_WAIT)) {
+		rep = NULL;
+	} else {
+		memset(rep, 0, sizeof(*rep));
+	}
+	return rep;
+}
+
+/**
+ * @brief Free the allocated memory for report
+ *
+ * Frees the memory allocated for report and stores NULL in the pointed
+ * report pointer.
+ *
+ * @param repp Pointer to the report.
+ */
+static void rep_free(struct bt_gatt_hids_c_rep_info **repp)
+{
+	k_mem_slab_free(&bt_gatt_hids_c_reports_mem, (void**)repp);
+	*repp = NULL;
+}
+
+/**
+ * @brief Create new report structure
+ *
+ * Allocate memory for new report and fill it with data from given report
+ * characteristic.
+ *
+ * @param hids_c   HIDS client object.
+ * @param repp     Pointer with would be set to newly created report.
+ * @param dm       Discovery object.
+ * @param rep_chrc Report characteristic.
+ *
+ * @return 0 or negative error code.
+ *
+ * @sa rep_alloc
+ * @sa rep_free
+ *
+ * @note If error is found during report initialization,
+ *       allocated report memory is freed.
+ */
+static int rep_new(struct bt_gatt_hids_c *hids_c,
+		   struct bt_gatt_hids_c_rep_info **repp,
+		   const struct bt_gatt_dm *dm,
+		   const struct bt_gatt_attr *rep_chrc)
+{
+	struct bt_gatt_hids_c_rep_info *rep;
+	const struct bt_gatt_attr *gatt_desc;
+	rep = rep_alloc();
+	if (!rep) {
+		LOG_ERR("No memory to create new report");
+		return -ENOMEM;
+	}
+	struct bt_gatt_chrc *chrc_val = bt_gatt_dm_attr_chrc_val(rep_chrc);
+	assert(chrc_val); /* This is internal function and it has to
```

Comment:
`__ASSERT_NO_MSG` for consistency with other files

Reply:
Fixed